### ISSUE TITLE
Question about handling received Detach after sending End

### ISSUE THREAD

**Issue by ChugR** [Azure/amqpnetlite#159]
I've been running the self tests against brokers other than the built in testbroker. When running against [Apache ActiveMQ Artemis](http://activemq.apache.org/artemis/download.html) several tests fail. I'd like to discuss one test as there are behaviors that might need adjustment in both the broker and the amqpnetlite client.

The test is TestMethod_LinkReopen. Over the wire the traffic is

```
Client                      Server
  | Open, Begin, Attach        |
  |--------------------------->|
  |                            |
  |        Open, Begin, Attach |
  |<---------------------------|
  |                            |
  | End, Close                 |
  |--------------------------->|
  |                            |
  |                     Detach | ???
  |<---------------------------|
  |                            |
  |                End, Close  |
  |<---------------------------|
  |                            |
```

The Lite client complains

```
[03:16:58.104] Exception occurred: Amqp.AmqpException: Operation 'Send' is not valid under state: CloseSent.
   at Amqp.Connection.ThrowIfClosed(String operation) in d:\Users\crolke\git\amqpnetlite\src\Connection.cs:line 390
   at Amqp.Connection.SendCommand(UInt16 channel, DescribedList command) in d:\Users\crolke\git\amqpnetlite\src\Connection.cs:line 244
   at Amqp.Session.SendCommand(DescribedList command) in d:\Users\crolke\git\amqpnetlite\src\Session.cs:line 259
   at Amqp.Link.SendDetach(Error error) in d:\Users\crolke\git\amqpnetlite\src\Link.cs:line 268
   at Amqp.Link.OnDetach(Detach detach) in d:\Users\crolke\git\amqpnetlite\src\Link.cs:line 159
   at Amqp.Session.OnDetach(Detach detach) in d:\Users\crolke\git\amqpnetlite\src\Session.cs:line 497
   at Amqp.Session.OnCommand(DescribedList command, ByteBuffer buffer) in d:\Users\crolke\git\amqpnetlite\src\Session.cs:line 331
   at Amqp.Connection.OnSessionCommand(UInt16 remoteChannel, DescribedList command, ByteBuffer buffer) in d:\Users\crolke\git\amqpnetlite\src\Connection.cs:line 533
   at Amqp.Connection.OnFrame(ByteBuffer buffer) in d:\Users\crolke\git\amqpnetlite\src\Connection.cs:line 622
[03:16:58.104] Conn error: error(condition:amqp:illegal-state,description:Operation 'Send' is not valid under state: CloseSent.)
```

The question is about the Detach frame that the broker returns. 
### Broker side
- Some other brokers such as Apache Qpid C++ broker do not return the Detach frame, just the End and Close. In those cases the test closes with no error. I have been suggesting that the Artemis just not send the Detach and all would be well.
### Client side
- The session and connection are on their way down and the client just slammed the link closed without a proper Detach. In this case the client could be more forgiving of receiving the Detach which in not really out of order.
### Is it a protocol error?

There is nothing in the AMQP spec preventing the broker from deciding to send a Detach at any time. There is an admittedly small window of time where the client could send an End when the broker sends a Detach and the two are received simultaneously by their peers.

Probably in the field this is not an issue. However, in the self tests it causes a couple of failures repeatedly every time.

**Comment by xinchen10** [comment]
@ChugR the client should handle the detach without throwing exception.

**Comment by xinchen10** [comment]
Fix available in v1.2.2.
